### PR TITLE
DAOS-11171 test: add dfuse mtime unit tests

### PR DESCRIPTION
**Pull request by daltonbohning**
Test-tag: dfuse,pr dfuse,daily_regression dfuse_unit

- Change dfuse time granularity to 1 (nanosecond)
- Add dfuse unit mtime unit tests
- Adjust dfuse_test to run with or without metadata tests
- Add metadata tests to ftest/daos_test/dfuse.py

Signed-off-by: Dalton Bohning <dalton.bohning@intel.com>

Required-githooks: true

### CODE REVIEW

File: src/tests/suite/dfuse_test.c
Comment (jolivier23) [inline src/tests/suite/dfuse_test.c:191]: given that it's timespec_gt would it be a bit clearer if it were

return t1.tv_nsec > t2.tv_nsec

**Review by jolivier23 — APPROVED** [review]

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c]: Agree. Fixed. Copy paste is evil...

**Review by wangshilong — DISMISSED** [review]

File: src/client/dfuse/dfuse_fuseops.c
Comment (ashleypittman) [inline src/client/dfuse/dfuse_fuseops.c:90]: This would have been set for a reason, although it probably dates from IOF.  I think it might have been to mask clock-skew between client and server and timestamps apparently going backwards however I will check this in the commit history and report my findings.

File: src/tests/ftest/daos_test/dfuse.py
Comment (ashleypittman) [inline src/tests/ftest/daos_test/dfuse.py:84]: The issue is with writeback cache where the kernel assumes control of mtime and size and from memory this test skips the interception tests when using writeback already.

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [inline src/tests/suite/dfuse_test.c:38]: If this is needed then I would prefer that all tests were enabled by default and if necessary then an option was used to disable some of them.  It would probably also be nicer if we still ran the "tests" but only reported the results to stdout rather than failing.

File: src/tests/ftest/daos_test/dfuse.py
Comment (daltonbohning) [reply src/tests/ftest/daos_test/dfuse.py:84]: If you run all variants, you can see almost all of the caching ones fail the mtime tests. I opened https://daosio.atlassian.net/browse/DAOS-11204 for this, but it was closed as duplicate / won't fix

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c:38]: Then we should probably model it after `daos_test`, which runs all tests by default, but has flags to optionally run specific tests. I think the concept of disabling rather than enabling becomes confusing as more tests are added.

> still ran the "tests" but only reported the results to stdout rather than failing

I think that would require a bit of refactoring.

Anyway, this test wasn't quite set up like some of the other unit tests, which is why I took this simple approach.

File: src/tests/ftest/daos_test/dfuse.py
Comment (ashleypittman) [reply src/tests/ftest/daos_test/dfuse.py:84]: That might have been a mistake, it was closed as a dupliate of a ticket that was closed because it was based on writeback caching.

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:38]: you probably have more experience of it than me, I'm just keen that we don't setup a situation where we can accidentally disable tests without noticing.  Happy to defer to you on this one.

Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c]: I overhauled the interface based on `dfs_test`. By default, all tests are ran, but each can be ran independently. `dfs_test` does have a `-E|--exclude` option, but it doesn't work, so I didn't include that.

**Comment by ashleypittman** [comment]
Merged this with master and I've enabled the metadata tests in all cases so we can see the results.

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [inline src/tests/suite/dfuse_test.c:218]: You replaced tabs with spaces? Is this our preferred style?

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [inline src/tests/suite/dfuse_test.c:289]: Tabs -> spaces?

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [inline src/tests/suite/dfuse_test.c:327]: FWIW, `daos_test` and `dfs_test` use the other style, which I think is cleaner

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:218]: I just ran git-clang-format on it, our policy on tabs is set at https://github.com/daos-stack/daos/blob/master/.clang-format#L10

**Comment by ashleypittman** [comment]
The test failures are a dfs issue, when setattr is called it first calls stat, then it applies the changes.  For truncate it does not then re-sample the remote mtime value.

https://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c#L4758-L4763

**Review by ashleypittman — DISMISSED** [review]

File: src/tests/suite/dfuse_test.c
Comment (mchaarawi) [inline src/tests/suite/dfuse_test.c:231]: should that be
assert_true(timespec_gt(now, stbuf.st_mtim)); ?

File: src/tests/suite/dfuse_test.c
Comment (mchaarawi) [inline src/tests/suite/dfuse_test.c:253]: i understand that 10 seconds is probably way more than enough time for the entire test to run, however in CI and especially with VM tests, we've seen some pretty weird behavior and some tests taking very long without any reason. so maybe this past time step should use the first timestamp on the file as a base instead of now.

File: src/tests/suite/dfuse_test.c
Comment (mchaarawi) [inline src/tests/suite/dfuse_test.c:263]: after manually setting the mtime in the previous step, it would be good to check that updating the file then stat, actually returns the more recent mtime.

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:231]: Yes, I think it probably should.

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:263]: This does that does't it?  Do you mean an extra check where we write to the file, then stat again and ensure that the file is newer than prev_ts?

Comment (mchaarawi) [reply src/tests/suite/dfuse_test.c:263]: yes indeed. here we set mtime with futimens. and verify that it's the one we set.
we just want to make sure if we touch the file again, that one we set is overwritten.
(just to verify an internal implementation detail). i did that in my dfs unit test, so it's not necessary to do here, but i see this test is mimicking the dfs one, so maybe we can just add it to make sure the dfuse level test behaves the same

**Review by mchaarawi — APPROVED** [review]

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c:218]: I still wonder if that clang format file is wrong, considering we use tabs everywhere else

**Review by daltonbohning — COMMENTED** [review]
LGTM. Thanks @ashleypittman

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:312]: When we looked at this there was a real mix with no clear style more common than another.

Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c:312]: Interesting... Anyway, not a blocker for me.

**Review by ashleypittman — APPROVED** [review]

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c:389]: @ashleypittman This is my doing, but I agree with Mohamad. It's more robust to use the first timestamp (or prev?) instead of `now`.

File: src/tests/suite/dfuse_test.c
Comment (daltonbohning) [reply src/tests/suite/dfuse_test.c:481]: Shouldn't we properly resolve this? FWIW, this style doesn't have that problem:
```
static struct option long_options[] = {
    {...}
}
```

**Review by daltonbohning — COMMENTED** [review]
LGTM - minor comments - can't approve my own PR

**Review by wangshilong — APPROVED** [review]

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:481]: Yes we should.  I'm thinking I'll push a PR to remove checkpatch warnings that conflict with our clang-format settings.

File: src/tests/suite/dfuse_test.c
Comment (ashleypittman) [reply src/tests/suite/dfuse_test.c:389]: We've got further work to do with this feature that is blocked on this so I'm going to land as-is and we can review if it becomes a problem.

**Pull request merged**
